rfcs/20200612-stream-executor-c-api.md

```diff
@@ -336,94 +401,145 @@ typedef struct SP_StreamExecutor {
+
+  void (*destroy_timer_fns)(const SP_Platform* platform,
+                            SP_TimerFns* timer_fns);
+} SP_PlatformFns;
+
+#define SP_PLATFORM_FNS_STRUCT_SIZE \
+  TF_OFFSET_OF_END(SP_PlatformFns, destroy_timer_fns)
 
 typedef struct SE_PlatformRegistrationParams {
   size_t struct_size;
-  void* ext;
-  
+  void* ext;  // reserved for future use
+
   // StreamExecutor C API version.
   int32_t major_version;
   int32_t minor_version;
   int32_t patch_version;
-  
-  // Must be filled by the plugin.
-  SP_Platform platform;  // out
+
+  SP_Platform* platform;         // output, set by plugin
+  SP_PlatformFns* platform_fns;  // output, set by plugin
+  // Clean up fields inside SP_Platform that were allocated
+  // by the plugin. `platform` itself should not be deleted here.
+  void (*destroy_platform)(SP_Platform* platform);  // out, set by plugin
+  void (*destroy_platform_fns)(
+      SP_PlatformFns* platform_fns);  // out, set by plugin
 } SE_PlatformRegistrationParams;
 
-#define SE_PLATFORM_REGISTRATION_PARAMS_SIZE TF_OFFSET_OF_END(SE_PlatformRegistrationParams, platform)
+#define SE_PLATFORM_REGISTRATION_PARAMS_STRUCT_SIZE \
+  TF_OFFSET_OF_END(SE_PlatformRegistrationParams, destroy_platform_fns)
 
-void SE_InitializePlugin(SE_PlatformRegistrationParams* params, TF_Status* status);
+void SE_InitPlugin(SE_PlatformRegistrationParams* params, TF_Status* status);
 
 #ifdef __cplusplus
-} // extern "C"
+}  // extern "C"
 #endif
 ```
 
-## Registration implementation
+### PlatformId
 
-Registration will be implemented by registering a new StreamExecutor platform as well as a new TensorFlow device with [DeviceFactory](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/common_runtime/device_factory.h;l=30?q=DeviceFactory).
+StreamExecutor [Platform](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/stream_executor/platform.h;l=114) has an id parameter. This parameter will be hidden from the C API and set
+internally by TensorFlow instead.
 
+
+## Usage Example
+Code example for [PluggableDevice](https://github.com/tensorflow/community/pull/262)
+registration outlined in the [Usage Overview](#Usage overview) section.
+
+### Core TensorFlow
 ```cpp
-typedef (*SEPluginInitFn)(SE_PlatformRegistrationParams*, TF_Status*);
+typedef void (*SEInitPluginFn)(SE_PlatformRegistrationParams*, TF_Status*);
 ...
 
-void* plugin = dlopen("myplugin.so", ...);
-if (!plugin) {
-  ... output error and skip this plugin ...
-}
-void* initialize_sym = dlsym(plugin, "SE_InitializePlugin");
+void* initialize_sym = dlsym(plugin_dso_handle, "SE_InitPlugin");
 if (!initialize_sym) {
-  ... output error and skip this plugin ...
+  // Output error and skip this plug-in.
 }
-SEPluginInitFn initialize_fn = reinterpret_cast<SEPluginInitFn>(initialize_sym);
+SEInitPluginFn initialize_fn = reinterpret_cast<SEInitPluginFn>(initialize_sym);
 
 SE_PlatformRegistrationParams params;
 TF_Status* status = TF_NewStatus();
```

Comment:
```suggestion
TF_Status status;

initialize_fn(&params, &status);
```

Reply:
`TF_Status* status = TF_NewStatus();` is correct. But `initialize_fn(&params, &status);` should instead be changed to `initialize_fn(&params, status);`

Reply:
I made that nit, that TF_NewStatus() seems to imply heap allocation as it returns a pointer (from another scope). This is an anti-pattern because the allocation can be avoided. As examples are often modeled after by others and sometimes copy-pasted, I think we should fix it.

Reply:
Sounds good.

Reply:
Thanks for accepting my feedback!

Reply:
Using `TF_Status status` sounds good. I was using the pointer form following the test file.

Unfortunately, this suggestion doesn't cover the existing ```initialize_fn(&params, status);``` line. So we will have two duplicate `initialize_fn` calls after this commit (that is purely from accepting suggested changes).  I'll edit the old one out in a separate commit.